### DISCUSSION TITLE
Why "/opt/homebrew" only?

### DISCUSSION THREAD

**Discussion by tjconcept** [Homebrew/discussions#440] [Everyday usage]
In the process of installing a new M1 machine, I've come across multiple versions of the "install into /opt/homebrew" recommendation (like https://github.com/Homebrew/brew/issues/7857#issuecomment-727561968), but none of them say *why*.

I'm curious what makes that path so special, and why Homebrew is dumping the "install anywhere" feature?

I've personally always avoided installing anything into non-user directories (having Homebrew at `~/Applications/homebrew`) and never noticed any issues.

**Comment by carlocab** [comment]
You can still install anywhere you want. However, a number of bottles (pre-built binary packages) will not work for you if you don't install in `/opt/homebrew`.

If you use a non-`/opt/homebrew` prefix, some of these formulae will just refuse to pour a bottle. You'll have to build those from source. Other formulae may still pour a bottle, but it's difficult to guarantee they'll work as well as if you used `/opt/homebrew` as your prefix.

**Reply by tjconcept** [reply]
Thanks!

But again.. What's the reason for this? I've not before encountered pre-built binaries or other application installations that would require a specific location.

Reading the comment (https://github.com/Homebrew/brew/issues/7857#issuecomment-727561968) it sounds like the option to install anywhere is deprecated.

**Reply by carlocab** [reply]
Some bottles have paths hard-coded into their binaries at build-time, and there's very little one code do about that without extensive patching of the source. Thus, when the bottle starts looking for external dependencies, it'll look for them in `/opt/homebrew/{bin,lib,include,*}`, because that's where they were at build time. That clearly won't work for you if you don't have anything in `/opt/homebrew`. (On Intel, you can say the same except it's `/usr/local` instead of `/opt/homebrew`.)

The status of install-anywhere is the same as it has been for a while. It wasn't supported before, and it's not going to start being supported now.

You can still do it if you want to, but Homebrew won't offer as much support when your Homebrew installation breaks.

See https://docs.brew.sh/Installation#untar-anywhere.

>However do yourself a favour and install to /usr/local on macOS Intel, /opt/homebrew on macOS ARM, and /home/linuxbrew/.linuxbrew on Linux. Some things may not build when installed elsewhere. One of the reasons Homebrew just works relative to the competition is because we recommend installing here. _Pick another prefix at your peril!_

**Reply by tjconcept** [reply]
> bottles have paths hard-coded into their binaries at build-time

That makes sense, but then those should just build from source, and would theoretically work just as well, right? Assuming the source didn't hardcode `/opt/homebrew` into their source code, but that sounds unlikely.

Summing up, the only reason is: some pre-built binaries may have paths hardcoded at build time and thus would need to be rebuild on installation.

I think it would be nice to have that on the website as well. When it says you're assuming a risk/doing a trade-off, it's nice to know what exactly that is 🙂

**Reply by carlocab** [reply]
>That makes sense, but then those should just build from source, and would theoretically work just as well, right?

Theoretically, yes. Note, however, that build systems are temperamental, and many things can cause builds to go wrong. Homebrew used to be a build-from-source package manager. We slowly made the switch to binary-packages-only largely because it was too hard to support users building from source. (Note that building from source is also not supported by Homebrew, largely for reasons already stated.)

>I think it would be nice to have that on the website as well.

Sure, maybe this would be useful to know. Perhaps you could open a PR to improve the docs?

**Reply by tjconcept** [reply]
I see. Thanks.

> Perhaps you could open a PR to improve the docs?

I'd love to, but I don't think I can actually contribute substantial value. It would be one of those only-half-on-point, symptom-treating-only, quality-and-consistency-reducing, time-consuming-to-argue-against PRs that are super easy to open, because the real work is in everything around it, and ends up stealing time from those adding value 😅

**Reply by gromgit** [reply]
> I'd love to, but I don't think I can actually contribute substantial value.

I'd argue that documentation improvement is _precisely_ the area that users like yourself can contribute usefully, and it's easily done right from the docs website. Everyone has blind spots, and Homebrew maintainers and veterans have been "deep in the weeds" so long that what's super-obvious to us can be confounding to everyone else.

**Reply by tjconcept** [reply]
I did look at the website, but there are just many more questions arising, which I can't answer, and which I think would be part of such.

First, I guess, with my limited knowledge, I would tackle this by adding some background information on how Homebrew works, what it does, and why it is like that. Obviously, I'm not in a position to write such.

The frontpage for example says:

> Homebrew won’t install files outside its prefix and you can place a Homebrew installation wherever you like.

I wouldn't say that at all, in light of the comment I've previously linked and the "it is not supported" stance. Thus, the "fix" might simply be to remove everything about this "install anywhere" if that's the direction anyway.

On a related note, the FAQ and documentation mention paths prefixed `~` multiple times, but that is, to me, contrary with the "won’t install files outside its prefix" statement, unless the `~` really means "the Homebrew folder".

**Reply by carlocab** [reply]
These are all good observations, @tjconcept. Now I think you really should open a PR. Just propose changes that you think will improve those pages. If you're not sure about something, just do what you think makes sense. What goes into your PR doesn't need to be a fully-polished end-product anyway -- that's what the review you get on your PR is for. (Of course, you should still try to make it good.)

**Reply by tjconcept** [reply]
> that you think will improve those pages

Well, that's the problem, I don't know how to improve them. The things I've mentioned are all things that puzzle me, but I have none of the answers, so I can't put anything in their place. I'm still not really wiser. The answers I've gotten so far, while probably correct, only contributed to more confusion.

My guess is that the contradictions and confusion is a reflection of diverging opinions or inconsistencies in the source code or between contributors, and thus these questions simply don't have answers.

**Reply by benjibee** [reply]
@carlocab I'm working on a PR now for the brew.sh site and I'd like to confirm a few assumptions before I edit the texts:

- This issue is dead-in-the-water, currently no plans exist to ever implement it.
- The documentation is incorrect in its language with statements like "you can place a Homebrew installation wherever you like"
- The various workarounds aren't supported and issues shouldn't be opened in Github is errors are encountered related to installing outside of the default prefixes.
- These workarounds shouldn't therefor be mentioned in the documentation, though stating they exist is acceptable.

One question I do have very much inline with @tjconcept's original question: when the switch to `/opt/homebrew` as a prefix was made how is this any different from allowing another custom prefix? Do you have any idea how many bottles simply hard-coded this new prefix for ARM installs or is it safe to assume that anything that works in ARM is using something akin to `$(brew --prefix)`?

Additionally, is there anything holding Homebrew back from asking all contributors to only use this to determine prefixes so bottles would eventually be mostly compatible with custom prefixes?

**Reply by SMillerDev** [reply]
> This issue is dead-in-the-water, currently no plans exist to ever implement it.

Implement what exactly?

> The documentation is incorrect in its language with statements like "you can place a Homebrew installation wherever you like"

You can, but the next line is also important "However do yourself a favour and use the installer to install to the default prefix. Some things may not build when installed elsewhere. One of the reasons Homebrew just works relative to the competition is because we recommend installing here. Pick another prefix at your peril!"

> when the switch to /opt/homebrew as a prefix was made how is this any different from allowing another custom prefix?

Because the location it's compiled with determines the hardcoded prefix, which on ARM is `/opt/Homebrew`. Changing it would require recompiling.

> Additionally, is there anything holding Homebrew back from asking all contributors to only use this to determine prefixes so bottles would eventually be mostly compatible with custom prefixes?

I'm not sure what you mean.

**Reply by benjibee** [reply]
> > This issue is dead-in-the-water, currently no plans exist to ever implement it.
> 
> Implement what exactly?
> 
Implement any sort of solution in Homebrew that would enable using custom prefixes in practice, not just in theory.
>
> > The documentation is incorrect in its language with statements like "you can place a Homebrew installation wherever you like"
> 
> You can, but the next line is also important "However do yourself a favour and use the installer to install to the default prefix. Some things may not build when installed elsewhere. One of the reasons Homebrew just works relative to the competition is because we recommend installing here. Pick another prefix at your peril!"
> 
This was in context of the previous conversation about submitting a PR updating this language to make it clearer. Saying you _could_ do a thing when it likely won't work and the project isn't heading in a direction that's pushing for support of it can be quite confusing.
>
> > when the switch to /opt/homebrew as a prefix was made how is this any different from allowing another custom prefix?
> 
> Because the location it's compiled with determines the hardcoded prefix, which on ARM is `/opt/Homebrew`. Changing it would require recompiling.
>
I guess that answers my question. Bottles for ARM have hard-coded prefixes.
> 
> > Additionally, is there anything holding Homebrew back from asking all contributors to only use this to determine prefixes so bottles would eventually be mostly compatible with custom prefixes?
> 
> I'm not sure what you mean.
>
I mean asking maintainers and contributors to use something akin to `$(brew --prefix)` to determine the prefix and avoid hard-coding it altogether.

**Reply by SMillerDev** [reply]
> I mean asking maintainers and contributors to use something akin to $(brew --prefix) to determine the prefix and avoid hard-coding it altogether.

I think you might be misunderstanding where most of hardcoding happens. Generally when homebrew builds something we tell the build where to find things like configuration files before compiling. That means you can't use that binary with a non-default prefix because then it would be looking at the CI/standard prefix instead of your prefix.

The only way around that is recompiling, which takes a lot of effort and might not always work with your prefix because of assumptions in the build for example.

**Reply by tjconcept** [reply]
One year in on my M1 with a custom path and I have yet to see any issues! ☺️

If someone hardcoded a path, I'd be tempted to trash their program as garbage 😏, but it hasn't happened.
